packages/core/src/tracing/idlespan.ts

```diff
@@ -0,0 +1,318 @@
+   * If this occurs we finish the transaction.
+   */
+  function _beat(): void {
+    // We should not be running heartbeat if the idle transaction is finished.
+    if (_finished) {
+      return;
+    }
+
+    const heartbeatString = Array.from(activities.keys()).join('');
+
+    if (heartbeatString === _prevHeartbeatString) {
+      _heartbeatCounter++;
+    } else {
+      _heartbeatCounter = 1;
+    }
+
+    _prevHeartbeatString = heartbeatString;
+
+    if (_heartbeatCounter >= 3) {
+      DEBUG_BUILD && logger.log('[Tracing] Transaction finished because of no change for 3 heart beats');
+      span.setStatus('deadline_exceeded');
+      _finishReason = FINISH_REASON_HEARTBEAT_FAILED;
+      span.end();
+    } else {
+      _pingHeartbeat();
+    }
+  }
+
+  /**
+   * Pings the heartbeat
+   */
+  function _pingHeartbeat(): void {
+    DEBUG_BUILD && logger.log(`pinging Heartbeat -> current counter: ${_heartbeatCounter}`);
+    setTimeout(() => {
+      _beat();
+    }, heartbeatInterval);
+  }
+
+  function endIdleSpan(): void {
+    const endTimestamp = span.endTimestamp;
+
+    // This should never happen, but to make TS happy...
+    if (!endTimestamp) {
+      return;
+    }
+
+    _finished = true;
+    activities.clear();
+
+    if (span.op === 'ui.action.click' && !span.tags[FINISH_REASON_TAG]) {
+      span.setTag(FINISH_REASON_TAG, _finishReason);
+    }
+
+    DEBUG_BUILD && logger.log('[Tracing] finishing idle span', new Date(endTimestamp * 1000).toISOString(), span.op);
+
+    childSpans.forEach(childSpan => {
+      // We cancel all pending spans with status "cancelled" to indicate the idle span was finished early
+      if (!childSpan.endTimestamp) {
+        childSpan.endTimestamp = endTimestamp;
+        childSpan.setStatus('cancelled');
+        DEBUG_BUILD &&
+          logger.log('[Tracing] cancelling span since span ended early', JSON.stringify(childSpan, undefined, 2));
+      }
+
+      const spanStartedBeforeIdleSpanEnd = childSpan.startTimestamp < endTimestamp;
+
+      // Add a delta with idle timeout so that we prevent false positives
+      const timeoutWithMarginOfError = (finalTimeout + idleTimeout) / 1000;
+      const spanEndedBeforeFinalTimeout = childSpan.endTimestamp - span.startTimestamp < timeoutWithMarginOfError;
+
+      if (DEBUG_BUILD) {
+        const stringifiedSpan = JSON.stringify(childSpan, undefined, 2);
+        if (!spanStartedBeforeIdleSpanEnd) {
+          logger.log('[Tracing] discarding Span since it happened after idle span was finished', stringifiedSpan);
+        } else if (!spanEndedBeforeFinalTimeout) {
+          logger.log('[Tracing] discarding Span since it finished after idle span final timeout', stringifiedSpan);
+        }
+      }
+    });
+
+    DEBUG_BUILD && logger.log('[Tracing] flushing idle span');
+
+    // Clear array of child spans
+    childSpans.splice(0, childSpans.length);
+  }
+
+  client.on('spanStart', span => {
+    if (_finished) {
```

Comment:
Instead of patching the span recorder, we hook into `spanStart` and check here if the new span is a (nested) child span of the idle span.